sympy/polys/rootoftools.py

```diff
@@ -29,22 +29,84 @@
 from sympy.polys.domains import QQ
 
 from mpmath import mpf, mpc, findroot, workprec
-from mpmath.libmp.libmpf import prec_to_dps
+from mpmath.libmp.libmpf import dps_to_prec, prec_to_dps
 
 from sympy.utilities import lambdify, public, sift
 
-from sympy.core.compatibility import range
+from sympy.core.compatibility import range, ordered
 
 from math import log as mathlog
 
 __all__ = ['CRootOf']
 
-def _ispow2(i):
-    v = mathlog(i, 2)
-    return v == int(v)
 
-_reals_cache = {}
-_complexes_cache = {}
+
+class _pure_key_dict(dict):
+    """A minimal dictionary that makes sure that the key is PurePoly.
+
+    Examples
+    ========
+
+    Only the following four actions are guaranteed:
+
+    >>> from sympy.polys.rootoftools import _pure_key_dict
+    >>> from sympy import S, PurePoly
+    >>> from sympy.abc import x, y
+
+    1) creation
+
+    >>> P = _pure_key_dict()
+
+    2) assignment for a PurePoly or univariate polynomial
+
+    >>> P[x] = 1
+    >>> P[PurePoly(x - y, x)] = 2
+
+    3) retrieval based on PurePoly key comparison (use this
+       instead of the get method)
+
+    >>> P[y]
+    1
+
+    4) KeyError when trying to retrieve a nonexisting key
+
+    >>> P[y + 1]
+    Traceback (most recent call last):
+    ...
+    KeyError: PurePoly(y + 1, y, domain='ZZ')
+    """
+    def __getitem__(self, k):
+        if not isinstance(k, PurePoly):
+            assert len(k.free_symbols) == 1
+            k = PurePoly(k, expand=False)
+        return self.__dict__[k]
```

Comment:
It seems that `__dict__` is not the place where a dict object stores its data.
```
>>> from sympy.polys.rootoftools import _pure_key_dict
>>> from sympy.abc import x
>>> P = _pure_key_dict()
>>> P[x] = 1
>>> x in P
False
>>> P.items()
[]
```
If we want to use other dict methods than `__setitem__` and `__getitem__`, we should probably either define them explicitly or else try to use here something like `dict.__getitem__(self, k)`.

Reply:
It is probably not necessary to make lightweight cache a subclass of dict, if its methods are not needed. It seems to me that there could be `self.cache = {}` in `__init__(self)` and `__setitem__(self, k, v)` could then use `self.cache[k] = v`.

Reply:
LOL. I meant "caveat emptor" not "laissez-faire"!